### PR TITLE
use docker manifest create inside beta push

### PR DESCRIPTION
**Pull request by Zoey2936**

### CODE REVIEW

**Comment by Zoey2936** [comment]
If it works this command: `curl -H "Authorization: Bearer $(curl "https://auth.docker.io/token?service=registry.docker.io&scope=repository:nextcloud/all-in-one:pull" -s | jq -r .token)" -H "Accept: application/vnd.docker.distribution.manifest.v2+json" https://registry-1.docker.io/v2/nextcloud/all-in-one/manifests/beta-arm64 -Is | grep docker-content-digest` should have an output

**Comment by Zoey2936** [comment]
or we make the arm64 tag also multiarch

**Comment by Zoey2936** [comment]
@szaimen what is your opinion?

**Comment by szaimen** [comment]
> @szaimen what is your opinion?

The PR looks good to me. Can you also do the same in the publish to latest workflow or did you want to do it in a follow-up?

**Comment by Zoey2936** [comment]
> > @szaimen what is your opinion?
> 
> The PR looks good to me. Can you also do the same in the publish to latest workflow or did you want to do it in a follow-up?

It won't work like this in the latest workflow, since manifest create works only with manifests and not manifest list currently (https://github.com/docker/cli/pull/3994)

**Comment by szaimen** [comment]
> only with manifests and not manifest list

Okay but I thought the goal was to retreive a `application/vnd.docker.distribution.manifest.v2+json` on beta-arm64 after this PR so that should be fine? Or do I misunderstand something?

**Comment by Zoey2936** [comment]
we build images with the `application/vnd.docker.distribution.manifest.v2+json` manifest
inside beta they get converted to `application/vnd.docker.distribution.manifest.v2+json` and `application/vnd.docker.distribution.manifest.list.v2+json`
and latest copies them

**Comment by szaimen** [comment]
Ah I see. So it should work even without touchin the publish to latest workflow?

**Comment by Zoey2936** [comment]
I think yes

**Comment by szaimen** [comment]
Okay, thanks! We'll see then :)

File: .github/workflows/promote-to-beta.yml [hunk 1]
```diff
@@ -46,9 +46,11 @@ jobs:
...
+            docker manifest create nextcloud/$AIO_NAME\:beta-arm64 -a nextcloud/$AIO_NAME\:develop-arm64
+            docker manifest push nextcloud/$AIO_NAME\:beta-arm64
             
         else
             # amd64
-            docker buildx imagetools create --tag nextcloud/$AIO_NAME\:beta nextcloud/$AIO_NAME\:develop
-        fi
\ No newline at end of file
+            docker manifest create nextcloud/$AIO_NAME\:beta -a nextcloud/$AIO_NAME\:develop
+            docker manifest push nextcloud/$AIO_NAME\:beta
```
Comment (szaimen) [inline .github/workflows/promote-to-beta.yml]: btw, why did you change this? Shouldn't only the beta-arm64 channel get change or will this not work?

File: .github/workflows/promote-to-beta.yml [hunk 2]
```diff
@@ -46,9 +46,11 @@ jobs:
             docker manifest push nextcloud/$AIO_NAME\:beta
             
             # arm64
-            docker buildx imagetools create --tag nextcloud/$AIO_NAME\:beta-arm64 nextcloud/$AIO_NAME\:develop-arm64
+            docker manifest create nextcloud/$AIO_NAME\:beta-arm64 -a nextcloud/$AIO_NAME\:develop-arm64
+            docker manifest push nextcloud/$AIO_NAME\:beta-arm64
```
Comment (szaimen) [inline .github/workflows/promote-to-beta.yml:50]: Isn't this enough to make it work?

File: .github/workflows/promote-to-beta.yml (hunk 1 above)
Comment (Zoey2936) [reply .github/workflows/promote-to-beta.yml]: I thought the workflow would then be more consistent

File: .github/workflows/promote-to-beta.yml (hunk 2 above)
Comment (Zoey2936) [reply .github/workflows/promote-to-beta.yml:50]: yes

File: .github/workflows/promote-to-beta.yml (hunk 1 above)
Comment (szaimen) [reply .github/workflows/promote-to-beta.yml]: I would like to keep this for now. I wouldn't touch this part since it seem to work currently.

File: .github/workflows/promote-to-beta.yml (hunk 2 above)
Comment (szaimen) [reply .github/workflows/promote-to-beta.yml:50]: okay cool!

File: .github/workflows/promote-to-beta.yml (hunk 1 above)
Comment (Zoey2936) [reply .github/workflows/promote-to-beta.yml]: reverted

File: .github/workflows/promote-to-beta.yml [hunk 3]
```diff
@@ -46,9 +46,10 @@ jobs:
...
```
Comment (szaimen) [inline .github/workflows/promote-to-beta.yml:50]: another question: based on this change, doesnt keeping the lines that I link mean that we are converting the manifests again?
https://github.com/nextcloud-releases/all-in-one/blob/eef7898f5c733b849ffe9ed2e88016f3afda9e9a/.github/workflows/promote-to-latest.yml#L63-L64

Comment (szaimen) [reply .github/workflows/promote-to-beta.yml:50]: So wouldn't we actually need to change it there as well?

Comment (Zoey2936) [reply .github/workflows/promote-to-beta.yml:50]: I would say no, since buildx then has a manifest and a manifets list as base. But would the savest way not be to build the arm64 tag also multiarch, even if it is called arm64? since multiarch works

Comment (szaimen) [reply .github/workflows/promote-to-beta.yml:50]: Basically we could then get rid of docker buildx imagetoosl, right?

Comment (Zoey2936) [reply .github/workflows/promote-to-beta.yml]: > Basically we could then get rid of docker buildx imagetoosl, right?

When https://github.com/docker/cli/pull/3994 is merged, released and available, we could use manifest create also inside the latest workflow, but I think there are no advantages

> > But would the savest way not be to build the arm64 tag also multiarch, even if it is called arm64? since multiarch works
> 
> Yeah, you have a point here.
> 
> Okay, would you change the workflows to implement this? :)

done

Comment (szaimen) [reply .github/workflows/promote-to-beta.yml]: > When https://github.com/docker/cli/pull/3994 is merged, released and available, we could use manifest create also inside the latest workflow, but I think there are no advantages

All right!

> done

Thanks! :)

**Review by szaimen — APPROVED** [review]

**Comment by Zoey2936** [comment]
seems to work

**Comment by szaimen** [comment]
Great 🎉

**Comment by Zoey2936** [comment]
But another question, what if a person migrates their AIO instance from arm64 to amd64, while keeping the compose file which contains the latest-amr64 tag, what will happen to clamav all other containers should work?

**Comment by szaimen** [comment]
> what will happen to clamav

clamav should then be shown as available since the internal logic relies on the used architecture:

https://github.com/nextcloud/all-in-one/blob/f1dff9b3e29fa8b6f9eadef00900027aaa478d2e/php/src/Data/ConfigurationManager.php#L128

**Comment by Zoey2936** [comment]
but there is no nextcloud/aio-clamav:latest-arm64 image? or would it pull nextcloud/aio-clamav:latest?

**Comment by szaimen** [comment]
> nextcloud/aio-clamav:latest-arm64

Ah right. You would need to switch to the correct tag in that case...

**Comment by szaimen** [comment]
But you would notice this since it would fail to download that tag since it doesnt exist.

**Comment by Zoey2936** [comment]
So the mastercontainer will throw errors?
I know that this is very unrealistic, but should there maybe be a logic that disallows clamav even if the `-arm64` tag is used, or should we push clamav also to the arm64 tag, even if it is amd64 only? Or are my thoughts unnecessary?

**Comment by szaimen** [comment]
> So the mastercontainer will throw errors?

The mastercontainer will indeed throw errors.

> I know that this is very unrealistic, but should there maybe be a logic that disallows clamav even if the `-arm64` tag is used, or should we push clamav also to the arm64 tag, even if it is amd64 only? Or are my thoughts unnecessary?

I hope that if people should experience this will open a ticket about this and then we should be able to help them pretty easily. I don't think this should happen often since hopefully people use the correct tag on the server instead of re-using the old one.
